account_payment_extension/migrations/8.0.1.0.0/pre-migration.py

```diff
@@ -0,0 +1,121 @@
+               'bank_type_payment_type_rel_old')
+    cr.execute('ALTER INDEX bank_type_payment_type_rel_pay_type_id_index '
+               'RENAME TO bank_type_payment_type_rel_pay_type_id_index_old')
+    cr.execute('ALTER INDEX bank_type_payment_type_rel_bank_type_id_index '
+               'RENAME TO bank_type_payment_type_rel_bank_type_id_index_old')
+    # Switch payment types by payment modes
+    cr.execute("SELECT id, name, active, company_id FROM payment_type")
+    payment_types = cr.fetchall()
+    for payment_type in payment_types:
+        if payment_type[3]:
+            applicable_companies_ids = [payment_type[3]]
+        else:
+            cr.execute("SELECT id FROM res_company")
+            applicable_companies_ids = [x[0] for x in cr.fetchall()]
+        for company_id in applicable_companies_ids:
+            # Update payment modes
+            cr.execute("""SELECT id, name FROM payment_mode
+                          WHERE type=%s""", (payment_type[0], ))
+            payment_modes = cr.fetchall()
+            for payment_mode in payment_modes:
+                openupgrade.logged_query(
+                    cr,
+                    """
+                    UPDATE payment_mode
+                    SET name=%s, active=%s
+                    WHERE id=%s""", (
+                        "%s (%s)" % (payment_type[1], payment_mode[1]),
+                        payment_type[2], payment_mode[0])
+                )
+            # Create at least one payment mode for each payment type
+            if not payment_modes:
+                # Search for a journal
+                cr.execute("""SELECT id FROM account_journal
+                              WHERE active=TRUE
+                              AND company_id=%s
+                              AND type='bank'
+                              LIMIT 1""",
+                           (company_id, ))
+                row = cr.fetchone()
+                journal_id = row and row[0] or None
+                # Search for a bank account
+                cr.execute("""SELECT rpb.id FROM res_partner_bank rpb
+                              INNER JOIN res_partner rp ON rpb.partner_id=rp.id
+                              INNER JOIN res_company rc ON rc.partner_id=rp.id
+                              WHERE rc.id=%s
+                              LIMIT 1""", (company_id, ))
+                row = cr.fetchone()
+                bank_id = row and row[0] or None
+                # Create the payment mode
+                openupgrade.logged_query(
+                    cr,
+                    """
+                    INSERT INTO payment_mode
+                    (%s, name, active, journal, bank_id, company_id)
+                    VALUES (%%s, %%s, %%s, %%s, %%s)
+                    RETURNING id;""" % openupgrade.get_legacy_name('type'),
+                    (payment_type[0], payment_type[1], payment_type[2],
+                     journal_id, bank_id, company_id))
```

Comment:
This pull works like a charm but I've have to change this insert code in order to be able to run it. For me works this:

```
                # Create the payment mode
                cr.execute(
                    """
                    INSERT INTO payment_mode
                    (id, name, active, journal, bank_id, company_id)
                    VALUES (%s, '%s', %s, %s, %s, %s)
                    RETURNING id;""" %
                    (payment_type[0], payment_type[1], payment_type[2],
                     journal_id, bank_id, company_id,))
```